runtime/complex.go

```diff
@@ -169,6 +178,185 @@ func complexCoerce(o *Object) (complex128, bool) {
 	return complex(floatO, 0.0), true
 }
 
+func complexNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc == 0 {
+		return newObject(t), nil
+	}
+	if argc > 2 {
+		return nil, f.RaiseType(TypeErrorType, "'__new__' of 'complex' requires at most 2 arguments")
+	}
+	if t != ComplexType {
+		// Allocate a plain complex then copy it's value into an object
+		// of the complex subtype.
+		x, raised := complexNew(f, ComplexType, args, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		result := toComplexUnsafe(newObject(t))
+		result.value = toComplexUnsafe(x).Value()
+		return result.ToObject(), nil
+	}
+	o := args[0]
+	if complexSlot := o.typ.slots.Complex; complexSlot != nil && argc == 1 {
+		c, raised := complexConvert(complexSlot, f, o)
+		if raised != nil {
+			return nil, raised
+		}
+		return c.ToObject(), nil
+	}
+	if o.isInstance(StrType) {
+		if argc == 2 {
+			return nil, f.RaiseType(TypeErrorType, "complex() can't take second arg if first is a string")
+		}
+		s := toStrUnsafe(o).Value()
+		result, err := parseComplex(s)
+		if err != nil {
+			return nil, f.RaiseType(ValueErrorType, "complex() arg is a malformed string")
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+	if argc == 2 && args[1].isInstance(StrType) {
+		return nil, f.RaiseType(TypeErrorType, "complex() second arg can't be a string")
+	}
+	var cr complex128
+	crIsComplex := false
+	if complexSlot := o.typ.slots.Complex; complexSlot != nil {
```

Comment:
Probably worth factoring this `if slots.Complex/else if slots.Float` piece out into a helper to eliminate some duplicate code.